### PR TITLE
Core: Add in RenameTableRequest for the RESTCatalog

### PR DESCRIPTION
**Pull request by kbendick**
Adds `RenameTableRequest` as a request for the `RESTCatalog` and adds the call for `renameTable` to the `RESTCatalog`.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/rest/requests/RenameTableRequest.java
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/rest/requests/RenameTableRequest.java:50]: Do we want to allow for the empty namespace to be used? Its not allowed in some other catalogs, but it seems like that might be beneficial for e.g. catalogs that don't support namespaces.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (kbendick) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:541]: This test is pretty basic as I didn't want to make the PR too large.

Let me know if I should add more `CatalogTests` for the `renameTable` operation in this PR 🙂

File: core/src/main/java/org/apache/iceberg/rest/requests/RenameTableRequest.java
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/requests/RenameTableRequest.java:50]: Yes, I don't see a reason not to allow requesting the empty namespace. That's valid in many catalogs.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:541]: I'd prefer adding tests for missing source table, existing destination table, and others. It's a good time to add them.

File: core/src/main/java/org/apache/iceberg/rest/requests/RenameTableRequest.java
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/requests/RenameTableRequest.java:50]: Ok will remove that.

Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/requests/RenameTableRequest.java]: Removed.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:587]: I added test for missing source and existing destination.

However, the JdbcCatalog does not throw the correct error for existing destination table, which can be verified by looking at the current JdbcCatalog tests: https://github.com/apache/iceberg/blob/62f3ca334c4bf377b1aaf9b7386bec1bfaddb5cd/core/src/test/java/org/apache/iceberg/jdbc/TestJdbcCatalog.java#L385-L390

I have put togeher a hacky fix to it in this PR, but I think we should just throw an `@Ignore` on this particular test and then fix the jdbc catalog in a separate PR. All of the inforrmation from the error was very SQLIte specific and I'm not sure of a general way to handle this. Open to suggestions.

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:198]: I don't think this test is going to hold up for most JDBC databases.

There's a SQLCode on the error, but it was SQLIte specific in the test runs.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:587]: I hit the same problem when testing create table atomicity and [fixed it similarly](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java#L126-L129). There's more background in [this thread](https://github.com/apache/iceberg/pull/4319#pullrequestreview-909504774).

I'd probably use the same logic as in that fix, which checks the message for "constraint failed".

File: core/src/main/java/org/apache/iceberg/rest/CatalogHandlers.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/CatalogHandlers.java:250]: I don't think other methods in `CatalogHandlers` run `validate`. I think that should be done before passing it in here.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:597]: `create` -> `rename`.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:640]: It would be nice to have a validation that the destination table is distinct from the source table. Maybe checking UUID or making it a partitioned table and ensuring it is still partitioned.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:644]: These get cleaned up with the catalog, so no need to drop and assert.

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:198]: We looked into this and there isn't a good way to do it. In `JdbcTableOperations`, we do the same thing but check for just "constraint failed". I pasted some links for background in my other comment.

**Review by rdblue — CHANGES_REQUESTED** [review]
@kbendick, this looks ready to go to me, except for the unfinished changes from the JDBC catalog issue. If you remove the comments in tests and update the `contains` check, I'll merge this.

**Comment by kbendick** [comment]
> @kbendick, this looks ready to go to me, except for the unfinished changes from the JDBC catalog issue. If you remove the comments in tests and update the `contains` check, I'll merge this.

Cool. I figured that the JDBC catalog issue was something there wasn't a clean way to resolve.

I will make the changes so we can merge this.

File: core/src/main/java/org/apache/iceberg/rest/CatalogHandlers.java
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/CatalogHandlers.java:250]: Oh good point they don't. Since we're building them via the builder, it's not really necessary. Plus validate is called in the constructor anyway.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:640]: I'll check the UUIDs. That seems simplest.

File: core/src/main/java/org/apache/iceberg/rest/requests/RenameTableRequest.java
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/requests/RenameTableRequest.java:50]: I suppose it can be a no-op. I looked at the Glue catalog, and it doesn't specifically check for this case.

In terms of making it a no-op, should we just let the client send the request and allow the server to deal with it?

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:198]: Updated to use the referenced code (with the same comment).

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:588]: Updated the check.

File: core/src/main/java/org/apache/iceberg/rest/requests/RenameTableRequest.java
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/requests/RenameTableRequest.java]: Allowing it to go through and the server can deal with it.

This is in line with some of the other catalogs.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:609]: Unnecessary comment?

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:139]: Given it's a POST request, the 200 won't be cached so I think either is fine.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:637]: Looks good.

**Review by rdblue — APPROVED** [review]
Looks great. I think there was a `WIP` comment accidentally left in, but overall looks good to merge.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:139]: Looking at the spec, I think that 204 might not be the best, as that indicates that the page should not be refreshed.

If a web UI were to receive a 204 response, it would possibly not refresh the page, but if the rename is successful and the page lists the current tables, then the UI should refresh.

https://www.rfc-editor.org/rfc/rfc2616#section-10.2.5

```
[10.2.5](https://www.rfc-editor.org/rfc/rfc2616#section-10.2.5) 204 No Content

   The server has fulfilled the request but does not need to return an
   entity-body, and might want to return updated metainformation. The
   response MAY include new or updated metainformation in the form of
   entity-headers, which if present SHOULD be associated with the
   requested variant.

   If the client is a user agent, it SHOULD NOT change its document view
   from that which caused the request to be sent. This response is
   primarily intended to allow input for actions to take place without
   causing a change to the user agent's active document view, although
   any new or updated metainformation SHOULD be applied to the document
   currently in the user agent's active view.

   The 204 response MUST NOT include a message-body, and thus is always
   terminated by the first empty line after the header fields.
```

That said, given that there's no response body, I am open either way.

Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:139]: The 200 content does say the POST body should have the updated information in it though, so I'm ok either way: https://www.rfc-editor.org/rfc/rfc2616#section-10.2.1

```
[10.2.1](https://www.rfc-editor.org/rfc/rfc2616#section-10.2.1) 200 OK

   The request has succeeded. The information returned with the response
   is dependent on the method used in the request, for example:

   GET    ....

   POST   an entity describing or containing the result of the action;
```

**Comment by kbendick** [comment]
> Looks great. I think there was a `WIP` comment accidentally left in, but overall looks good to merge.

Just noticed it. Removed!

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:139]: I'm good with 200.

**Comment by kbendick** [comment]
This should be good to go when you get a chance. Rebased and removed unnecessary commits.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java]: Removed!

**Review by rdblue — APPROVED** [review]

**Pull request merged**
